internal/httpproxy/cert_reloader.go

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) Twingate Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package httpproxy
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+type certReloader struct {
+	certFile string
+	keyFile  string
+	logger   *zap.Logger
+
+	watcher *fsnotify.Watcher
+	cancel  context.CancelFunc
+
+	mu   sync.RWMutex
+	cert *tls.Certificate
+}
+
+func newCertReloader(certFile, keyFile string, logger *zap.Logger) *certReloader {
+	return &certReloader{
+		certFile: certFile,
+		keyFile:  keyFile,
+		logger:   logger,
+	}
+}
+
+func (cr *certReloader) run(ctx context.Context) {
+	ctx, cr.cancel = context.WithCancel(ctx)
+
+	go wait.Until(func() {
+		if err := cr.watch(ctx); err != nil {
+			cr.logger.Error("failed to watch cert and key file, will retry later", zap.Error(err))
+		}
+	}, time.Minute, ctx.Done())
+}
+
+func (cr *certReloader) load() error {
+	cert, err := tls.LoadX509KeyPair(cr.certFile, cr.keyFile)
+	if err != nil {
+		return err
+	}
+
+	cr.mu.Lock()
+	defer cr.mu.Unlock()
+
+	cr.cert = &cert
+
+	return nil
+}
+
+func (cr *certReloader) watch(ctx context.Context) error {
+	var err error
+
+	cr.watcher, err = fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("error creating fsnotify watcher: %w", err)
+	}
+	defer cr.watcher.Close()
+
+	if err := cr.watcher.Add(cr.certFile); err != nil {
+		return fmt.Errorf("error adding watch for file %s: %w", cr.certFile, err)
+	}
+
+	if err := cr.watcher.Add(cr.keyFile); err != nil {
+		return fmt.Errorf("error adding watch for file %s: %w", cr.keyFile, err)
+	}
+
+	if err := cr.load(); err != nil {
+		return fmt.Errorf("error loading certificate: %w", err)
+	}
+
+	cr.logger.Info("Start watching cert and key files changes")
+
+	for {
+		select {
+		case event := <-cr.watcher.Events:
+			if err := cr.handleWatchEvent(event); err != nil {
+				return err
+			}
+		case err := <-cr.watcher.Errors:
+			cr.logger.Error("received error from watcher", zap.Error(err))
+
+			return fmt.Errorf("received error from watcher: %w", err)
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (cr *certReloader) handleWatchEvent(event fsnotify.Event) error {
+	cr.logger.Debug("Received watch event", zap.Any("event", event))
+
+	if !event.Has(fsnotify.Remove) && !event.Has(fsnotify.Rename) {
+		if err := cr.load(); err != nil {
+			cr.logger.Error("failed to load cert or key file", zap.Error(err))
+
+			return nil
+		}
+
+		cr.logger.Info("reloaded cert and key files")
+
+		return nil
+	}
+
+	if err := cr.watcher.Remove(event.Name); err != nil {
+		cr.logger.Info("failed to remove file watch, it may have been deleted", zap.Error(err))
+	}
+
+	if err := cr.watcher.Add(event.Name); err != nil {
+		cr.logger.Error("error adding watch for file", zap.String("filename", event.Name), zap.Error(err))
```

Comment:
does this error need to be logged, or will it be fine to let the `err` be logged by the caller?